lib/utils/httputil/httputil.go

```diff
@@ -314,18 +320,25 @@ func Send(method, rawurl string, options ...SendOption) (resp *http.Response, er
 		// TODO (@evelynl): disable retry after tls migration.
 		if err != nil && req.URL.Scheme == "https" && !httpFallbackDisabled {
 			log.Warnf("Failed to send https request: %s. Retrying with http...", err)
-			var httpReq *http.Request
-			httpReq, err = newRequest(method, opts)
+			originalErr := err
+			resp, err = fallbackToHTTP(client, method, opts)
 			if err != nil {
-				return nil, err
+				// Sometimes the request fails for a reason unrelated to https.
+				// To keep this reason visible, we always include the original
+				// error.
+				err = fmt.Errorf(
+					"failed to fallback https to http, original https error: %s,\n"+
+						"fallback http error: %s", originalErr, err)
 			}
-			httpReq.URL.Scheme = "http"
-			resp, err = client.Do(httpReq)
-		}
-		if err != nil {
-			continue
 		}
-		if resp.StatusCode >= 500 && !opts.acceptedCodes[resp.StatusCode] {
+		if err != nil ||
+			(isRetryable(resp.StatusCode) && !opts.acceptedCodes[resp.StatusCode]) ||
+			(opts.retry.extraCodes[resp.StatusCode]) {
```

Comment:
Can you wrap `(isRetryable(resp.StatusCode) && !opts.acceptedCodes[resp.StatusCode]) ||
			(opts.retry.extraCodes[resp.StatusCode])` in a function? so it becomes a bit more readable